armory/instrument/instrument.py

```diff
@@ -0,0 +1,670 @@
+    def _write(self, name, batch, result):
+        self.records.append((name, batch, result))
+
+    def collate_results(self):
+        """
+        Return a map from name to output, in original order.
+        """
+        output = {}
+        for name, batch, result in self.records:
+            if name not in output:
+                output[name] = []
+            output[name].append(result)
+        return output
+
+    def close(self):
+        output = self.collate_results()
+        with open(self.filepath, "w") as f:
+            json.dump(output, f)
+        # TODO: push to results dictionary?
+
+
+# GLOBAL CONTEXT METHODS #
+
+
+def get_context():
+    """
+    Get the context state object for the experimental procedure
+    """
+    global _CONTEXT
+    if _CONTEXT is None:
+        _CONTEXT = Context()
+    return _CONTEXT
+
+
+def get_probe(name: str = ""):
+    """
+    Get a probe with specified name, creating it if needed
+    """
+    if name != "" and not str.isidentifier(name):
+        raise ValueError(f"name {name} should be an identifier or the empty string")
+
+    if name not in _PROBES:
+        probe = Probe(name)
+        probe.set_sink(get_context())
+        _PROBES[name] = probe
+    return _PROBES[name]
+
+
+def connect_meter(meter):
+    get_context().connect_meter(meter)
+
+
+def add_meter(*args, **kwargs):
+    meter = Meter(*args, **kwargs)
+    connect_meter(meter)
+
+
+def get_meters():
+    return get_context().get_meters()
+
+
+def add_writer(writer):
+    get_context().add_writer(writer)
+
+
+def main():
+    """
+    Just for current WIP demonstration
+    """
+    # Begin model file
+    import numpy as np
+
+    # from armory.instrument import get_probe
+    probe = get_probe("model")
+
+    class Model:
+        def __init__(self, input_dim=100, classes=10):
+            self.input_dim = input_dim
+            self.classes = classes
+            self.preprocessor = np.random.random(self.input_dim)
+            self.predictor = np.random.random((self.input_dim, self.classes))
+
+        def predict(self, x):
+            x_prep = self.preprocessor * x
+            # if pytorch Tensor: probe.update(lambda x: x.detach().cpu().numpy(), prep_output=x_prep)
+            probe.update(lambda x: np.expand_dims(x, 0), prep_output=x_prep)
+            logits = np.dot(self.predictor.transpose(), x_prep)
+            return logits
+
+    # End model file
+
+    # Begin metric setup (could happen anywhere)
+
+    # from armory.instrument import add_writer, add_meter
+    from armory.utils import metrics
+
+    add_writer(PrintWriter())
+    add_meter(
+        "postprocessed_l2_distance",
+        metrics.l2,
+        "model.prep_output[benign]",
+        "model.prep_output[adversarial]",
+    )
+    add_meter(  # TODO: enable adding context for iteration number of attack
+        "sum of x_adv", np.sum, "attack.x_adv",  # could also do "attack.x_adv[attack]"
+    )
+    add_meter(
+        "categorical_accuracy",
```

Comment:
In general, the passing of values to `add_meter()` via strings with `.` in the middle I'm finding a bit non-intuitive. 

E.g. where `"scenario.y"` is being passed, I originally assumed this was somehow using the attribute of a scenario object, but only later realized the `scenario` comes from the probe name and `y` the kwarg passed to `update()`. Perhaps this is a non-issue, but it initially confused me

Reply:
I structured this so that if you wanted to, you could use the python namespace for this and ensure no collisions. E.g., `get_probe(__name__)`. Since However, if I want to guarantee no name collisions, I need to check for `'.'` in update. I will add that now, but we can discuss the broader naming convention later.